core/templates/services/playthrough.service.ts

```diff
@@ -57,96 +57,128 @@ class MultipleIncorrectAnswersTracker {
-   * visited states.
-   *
-   * If appending the new state would introduce a duplicate state name, then a
-   * cycle has been found. Specifically, if pushing the new state `N` onto the
-   * path would result in the following pattern:
-   *
-   *    [ ..., N , ... ] => [ ..., N , ..., N ]
-   *
-   * ...then we update this tracker's latest cycle discovery.
-   *
-   * The cycle is defined as the current path of visited states with all states
-   * prior to the first occurrence of N discarded.
-   *
-   * If this *exact* cycle has been discovered before (NOTE: rotations of a
-   * cycle are considered to be different from each other), then we increase the
-   * tracked number of cycle occurrences.
-   * Otherwise, the tracker is completely reset to 1.
-   *
-   * Finally, the path of visited states is reset to a value of [ N ], in hopes
-   * that the exact same cycle is built enough times to be considered an issue.
+   * Returns whether the two given cycles are the same sequence of state names,
+   * even when their starting element is different.
    *
-   * Note that 1-cycles (N -> N) are exceptional and ignored completely. They do
-   * not reset the tracker and do not increment the tracker's cycle occurrences.
+   * We check this by verifying whether the cycles can be rotated to become
+   * another. An array rotation is an operation which moves N elements from one
+   * end of the array to the other. For example:
+   *     RotateRight([1, 2, 3, 4, 5], 2) => [4, 5, 1, 2, 3].
+   *     RotateLeft([1, 2, 3, 4, 5], 2) => [3, 4, 5, 1, 2].
    */
-  recordStateTransition(destStateName: string): void {
-    if (this.currStateName() === destStateName) {
-      return;
+  private isSameCycleOfStateNames(
+      cycle: string[], cycleToCheck: string[]): boolean {
+    if (cycle.length !== cycleToCheck.length) {
+      return false;
     }
-    if (!this.pathOfVisitedStates.includes(destStateName)) {
-      this.pathOfVisitedStates.push(destStateName);
-    } else {
-      const cycleOfVisitedStates = (
-        this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
-      if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
-        this.cycleOccurrences += 1;
-      } else {
-        this.cycleOfVisitedStates = cycleOfVisitedStates;
-        this.cycleOccurrences = 1;
+    const rotationToFind = cycleToCheck.slice(1);
+    const arrayWithAllCycleRotations = [...cycle.slice(1), ...cycle.slice(1)];
+    const startIndex = arrayWithAllCycleRotations.indexOf(rotationToFind[0]);
+    return startIndex !== -1 && rotationToFind.every(
+      (state, i) => arrayWithAllCycleRotations[startIndex + i] === state);
+  }
+
+  private getMostCommonCycle(): [string[], number] {
+    let mostCommonCycle = null;
+    let mostCommonCycleOccurrences = 0;
+    for (const [cycle, occurrences] of this.cycleOccurrences.entries()) {
+      // When cycles have equal occurrences, take the most recent one.
```

Comment:
Oh -- you should probably leave a comment above the definition of cycleOccurrences explaining the sort order.

(Is there a sort order though? It's a map, which I thought didn't have one. But in that case there should be some kind of recency check here ... no?)

(I know in the past that in JS, the ordering for dicts is actually undefined, though you can often predict it for specific browsers. From a quick search, Maps seem to sort by recency so this probably works, though you'll still want to leave a note there and ensure good test coverage for this behaviour -- maybe a case of 50 cycles with 1 occurrence each -- so that no one calls sort() on it.)

Reply:
Maps are kept in insertion order, will clarify with a comment.